### PR TITLE
Create DSP Control Code Sample

### PR DESCRIPTION
**Pull request by shuoniu-intel**
## Description
Create new code sample to demonstrate how to set implementation preference for math functions, which affects hardware resource utilization on FPGA device.
## Testing
- [x] Linux: manual test (command line compile + HW run)
- [x] Linux: regtest (command line compile + report check + HW run)
- [ ] Windows: manual test (command line compile) (Currently blocked by [HSDES: 14015723889](https://hsdes.intel.com/appstore/article/#/14015723889))
- [ ] Windows: regtest (command line compile + report check + VS compile) (Currently blocked by [HSDES: 14015723889](https://hsdes.intel.com/appstore/article/#/14015723889))

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04* 
+| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit 
+| What you will learn               |  How to apply global DSP control in command-line interface. <br> How to apply local DSP control in source code. <br> Scope of datatypes and math operations that support DSP control.
+| Time to complete                  | 15 minutes
+
+## Purpose
+This tutorial shows how to apply global and local DSP controls to set implementation preference for math operations to control usage of DSPs. The global control is applied via command-line option and affects math operations in all kernels. The local control is applied via library function and affects math operations in a block scope in one single kernel. Both global and local controls only affect math operations that support DSP control.
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: This tutorial shows how to apply global and local ~~DSP~~ controls to set *the* implementation preference *between DSPs and soft-logic* for math operations ~~to control usage of DSPs~~. The global control is applied ~~via~~ *using a* command-line ~~option~~ *flag* and affects *applicable* math operations in all kernels. The local control is applied ~~via~~ *as a* library function and affects math operations in a block scope in ~~one~~ *a* single kernel. Both global and local controls only affect math operations that support DSP control.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+| `ac_int`               | Constant mul
+| `ac_fixed`             | Constant mul
+
+**NOTE:** _constant mul_ means one operand of the multiplication is a constant.
+
+### Global Control
+The `-Xsdsp-mode` command-line option sets implementation preference of math operations that support DSP control in all kernels. It has three valid options:
+| Option                           | Explanation
+---                                |---
+| `-Xsdsp-mode=default`            | This is the default. The compiler determines the implementation based on datatype and math operation.
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: I don't think you need to repeat `-Xsdsp-mode=` here if you add `-Xsdsp-mode=<option>` to the paragraph above.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+| `-Xsdsp-mode=prefer-dsp`         | Prefer math operations to be implemented in **DSPs**. If a math operation is implemented by DSPs by default, you will see no difference in resource utilization or area. Otherwise, you will notice decrease in the usage of soft-logic resources and increase in the usage of DSPs.
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: Otherwise, you will notice *a* decrease in the usage of soft-logic resources and *an* increase in the usage of DSPs.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+| `-Xsdsp-mode=prefer-softlogic`   | Prefer math operations to be implemented in **soft-logic**. If a math operation is implemented without DSPs by default, you will see no difference in resource utilization or area. Otherwise, you will notice decrease in the usage of DSPs and increase in the suage of soft-logic resources.
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: Otherwise, you will notice *a* decrease in the usage of DSPs and *an* increase in the ~~suage~~ usage of soft-logic resources.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md:35]: Make sure you run this README through a spell checker.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+### Local Control
+The library function `math_dsp_control<Preference::<enum>, Propagate::<bool>>([&]{})` provides block scope local control in one single kernel. A reference-capturing lambda expression is passed as the argument to this library function. Inside the lambda expression, implementation preference of math operations that support DSP control will be determined by two template arguments.
+
+The first template argument `Preference` is an enum with three valid options:
+| Option                           | Explanation
+---                                |---
+| `Preference::DSP`                | This is the default. Prefer math operations to be implemented in **DSPs**. Its behavior on a math operation is equivalent to global control `-Xsdsp-mode=prefer-dsp`.
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: > This is the default

This is a bit confusing since the compiler defaults is the third. Can you rephrase this instad?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+| `Preference::Softlogic`          | Prefer math operations to be implemented in **soft-logic**. Its behavior on a math operation is equivalent to global control `-Xsdsp-mode=prefer-softlogic`.
+| `Preference::Compiler_default`   | Compiler determines the implementation based on datatype and math operation. Its behavir on a math operation is equivalent to global control `-Xsdsp-mode=default`.
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: Typo on the word "behavior".

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+The second template argument `Propagate` is a boolean that determines propagation of the first argument `Preference` to functions called in the lambda expression:
+| Option  | Explanation
+---       |---
+| `On`    | This is the default. `Preference` is propagated to all functions called in the lambda expression, and affects all math operations on the call graph in the lambda expression until a nested `math_dsp_control<>()` call stops the propatation.
+| `Off`   | `Preference` is not propagated to functions called in the lambda expression. So `Preference` only applies to math operations directly used in the lambda expression.
+
```
Comment (ajaykumarkannan) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md:51]: You may also want to show that you can call the function with no template params and you'll get the DSP option with propagate. A couple of code snippets could be useful to have in the README as well.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md (hunk 5 above)
Comment (shuoniu-intel) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md:35]: Done checking with a spell checker. Spotted and fixed some other typos.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md (hunk 8 above)
Comment (shuoniu-intel) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md:51]: Added a code snippet.

**Comment by shuoniu-intel** [comment]
Adding @mdbtucker for review even though testing is still blocked by the ongoing xmain-integration.
The only test done by now is manually doing CMake compile on Linux.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
+# DSP Control
+This FPGA tutorial demonstrates how to set implementation preference for math operations, which affects hardware resource utilization on FPGA device.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: This FPGA tutorial demonstrates how to set the implementation preference for certain math operations (add, subtract, and multiplication by a constant) between hardened DSP blocks and soft logic.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit 
+| What you will learn               |  How to apply global DSP control in command-line interface. <br> How to apply local DSP control in source code. <br> Scope of datatypes and math operations that support DSP control.
+| Time to complete                  | 15 minutes
+
+## Purpose
+This tutorial shows how to apply global and local controls to set the implementation preference between DSPs and soft-logic for math operations. The global control is applied using a command-line flag and affects applicable math operations in all kernels. The local control is applied as a library function and affects math operations in a block scope in a single kernel. Both global and local controls only affect math operations that support DSP control.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: ...for CERTAIN math operations.

... that support DSP control (SEE TABLE BELOW).

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+### Scope of Datatypes and math operations
+| Datatypes              | math operations
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: Datatype (singular, not datatypes)
Capitalize Math Operations, and call it 'Controllable Math Operations'

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+---                      |---
+| `float`                | Add, sub, constant mul
+| `hls_float<8, 23>`     | Add, sub, constant mul
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: What is an hls_float?  Is it supported in SYCL?  What are the templating arguments for?  Is it only specifically 8,23 that can be controlled, or any hls_float?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+| `int`                  | Constant mul
+| `ac_int`               | Constant mul
+| `ac_fixed`             | Constant mul
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: Update the entire table to get rid of the abbreviations.  Use:

addition, subtraction, multiplication by a constant

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+**NOTE:** _constant mul_ means one operand of the multiplication is a constant.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: Delete this, 'multiplication by a constant' makes this redundant

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+| Option  | Explanation
+---       |---
+| `On`    | `Preference` is propagated to all functions called in the lambda expression, and affects all math operations on the call graph in the lambda expression until a nested `math_dsp_control<>()` call stops the propagation. <br> **NOTE:** This option will be automatically applied if the template argument `Propagate` is not specified manually.
+| `Off`   | `Preference` is not propagated to functions called in the lambda expression. So `Preference` only applies to math operations directly used in the lambda expression.
+
+**NOTE:**
+1. `Preference` never applies to nested `math_dsp_control<>()` calls. Each nested `math_dsp_control<>()` has its own `Preference`.
+2. Local control overrides global control on a controlled math operation.
+
+You may omit both template arguments if you want to get the DSP option with propagate. For example:
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: This sentence doesn't make sense to me, what does 'get the DSP option with propogate' mean?

You've already stated what the default template values are above, I think you can delete this line and the following cpp code

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+```cpp
+ext::intel::math_dsp_control<>([&] { ... });
+// This equals to ext::intel::math_dsp_control<ext::intel::Preference::DSP, ext::intel::Propagate::On>([&] { ... });
+```
+
+## Key Concepts
+* How to apply global DSP control in command-line interface
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: ...control FROM THE COMMAND-LINE

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json [hunk 17]
```diff
@@ -0,0 +1,63 @@
+{
+  "guid": "77006CE8-2935-41C7-BA30-DF6EA07CD73A",
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json:2]: Confirm you generated a unique GUID?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+  "targetDevice": ["FPGA"],
+  "builder": ["ide", "cmake"],
+  "languages": [{"cpp":{}}],
+  "ciTests": {
+    "linux": [
+      {
+        "id": "fpga_emu",
+        "steps": [
+          "dpcpp --version",
+          "cd ../../..",
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json]: Linux has no long path issues, no need to do this.  Please only do the directory messing about in Windows.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+          "cmake ../Tutorials/Features/dsp_control",
+          "make fpga_emu",
+          "./dsp_control.fpga_emu"
+        ]
+      },
+      {
+        "id": "report",
+        "steps": [
+          "dpcpp --version",
+          "cd ../../..",
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json]: Don't do this in Linux

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+
+The first template argument `Preference::<option>` is an enum with three valid options:
+| Option               | Explanation
+---                    |---
+| `DSP`                | Prefer math operations to be implemented in **DSPs**. Its behavior on a math operation is equivalent to global control `-Xsdsp-mode=prefer-dsp`. <br> **NOTE:** This option will be automatically applied if the template argument `Preference` is not specified manually.
+| `Softlogic`          | Prefer math operations to be implemented in **soft-logic**. Its behavior on a math operation is equivalent to global control `-Xsdsp-mode=prefer-softlogic`.
+| `Compiler_default`   | Compiler determines the implementation based on datatype and math operation. Its behavior on a math operation is equivalent to global control `-Xsdsp-mode=default`.
+
+The second template argument `Propagate::<option>` is an enum that determines propagation of the first argument `Preference` to functions called in the lambda expression:
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: This tutorial does not explain this concept in a way that I understand.  Perhaps the sample code needs to demonstrate the use of this Propogate argument?  After reading this, I don't get it.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json (hunk 17 above)
Comment (shuoniu-intel) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/sample.json:2]: Confirmed, this is a unique GUID I generated.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md (hunk 20 above)
Comment (shuoniu-intel) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: I'll rephrase the README a bit, and also add use of this Propagate argument in the sample code.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md (hunk 12 above)
Comment (shuoniu-intel) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/dsp_control/README.md]: It should be [ap_float](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide/top/optimize-your-design/resource-use/data-types-and-operations/var-prec-fp-sup/declare-and-use-the-ac-data-types/declare-the-ap-float-data-type.html). It is only ap_float<8, 23> that can be controlled.

**Review by mdbtucker — APPROVED** [review]
Looks good, thanks for the updates Shuo.  Please close this PR and create a new PR targeting the main repo master branch.
